### PR TITLE
`FSDPConfig`

### PR DESCRIPTION
**Pull request by salmanmohammadi**
## Summary by CodeRabbit

- New Features
  - Introduced a dedicated, typed FSDP configuration with clear options (e.g., activation checkpointing, parameter offloading, state dict types, auto-wrap policies, mixed precision).
  - Improved validation by using structured fields for safer, clearer config handling.

- Bug Fixes
  - Normalization/migration now consistently renames FSDP keys and removes legacy/conflicting entries, improving reliability across versioned configs.

- Tests
  - Expanded coverage to validate FSDP config migration, renamed fields, and preservation of non-versioned settings.

### CODE REVIEW

File: src/axolotl/utils/schemas/config.py [hunk 1]
```diff
@@ -667,8 +668,7 @@ class AxolotlInputConfig(
         json_schema_extra={"description": "FSDP configuration"},
         deprecated="Configuring FSDP using `fsdp` is deprecated. Please use `fsdp_config` instead. ",
     )
-    # TODO @SalmanMohammadi strongly type this as its own schema
-    fsdp_config: dict[str, Any] | None = Field(
+    fsdp_config: FSDPConfig | None = Field(
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/config.py:671]: Would this mean we're now dropping fsdp1 support? Given that the fsdp_config schema only has fsdp2 configs, any prior fsdp1 configs would be dropped.

Comment (salmanmohammadi) [reply src/axolotl/utils/schemas/config.py:671]: `fsdp_config` should also include FSDP1 fields, am I missing some?

Comment (salmanmohammadi) [reply src/axolotl/utils/schemas/config.py:671]: There is also the old way of configuring FSDP1 through `cfg.fsdp` which remains.

Comment (NanoCode012) [reply src/axolotl/utils/schemas/config.py:671]: For ex:
```
fsdp_config:
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: MistralDecoderLayer
  fsdp_activation_checkpointing: true
```
With the new pydantic schema, would these be ignored?

Comment (salmanmohammadi) [reply src/axolotl/utils/schemas/config.py:671]: No, we strip `fsdp_` from any fields which have it and emit a deprecation warning to encourage users to just use `state_dict_type` etc. But those fields are common to both FSDP1 and FSDP2.

Comment (salmanmohammadi) [reply src/axolotl/utils/schemas/config.py:671]: There are tests which cover this in `test_normalize_config.py`.

Comment (NanoCode012) [reply src/axolotl/utils/schemas/config.py:671]: Thanks for clarifying. Forgot it gets stripped before the pydantic

**Review by NanoCode012 — APPROVED** [review]

**Comment by salmanmohammadi** [comment]
Please hold off on merging until #3167 lands.
